packages/flutter_tools/lib/src/android/gradle.dart

```diff
@@ -368,13 +410,44 @@ Future<void> _buildGradleProjectV1(FlutterProject project, String gradle) async
   printStatus('Built ${fs.path.relative(project.android.gradleAppOutV1File.path)}.');
 }
 
-Future<void> _buildGradleProjectV2(
+Future<void> _buildGradleProjectV3(
   FlutterProject flutterProject,
   String gradle,
   BuildInfo buildInfo,
   String target,
   bool isBuildingBundle,
 ) async {
+  final String buildGradle = await flutterProject.android.rootBuildGradle.readAsString();
+  if (buildGradle.contains(_kSubprojectsNode)) {
+    printStatus('Updating build.gradle for new plugin format...');
+    final String updatedBuildGradle = buildGradle.replaceAll(_kSubprojectsNode, '');
+    await flutterProject.android.rootBuildGradle.writeAsString(updatedBuildGradle);
+  }
+  final String settingsGradle = await flutterProject.android.gradleSettingsFile.readAsString();
+  if (settingsGradle.contains(_kLegacySettingsGradlePluginLogic) || settingsGradle.contains(_kAlternateLegacySettingsGradlePluginLogic)) {
```

Comment:
It looks like if there's no match here because they've tweaked their source slightly this falls through to `v2`, but then v2 has been edited to rely on there being plugin AARs to build and no subprojects. Am I reading that right? If so we probably want to warn/fail if we can't find this block but still detect subprojects. Or branch instead of mutate the existing logic and only go into the new logic when we're absolutely sure it's safe to. Otherwise I think we risk breaking strange edge cases.

Reply:
So basically grep for `'subproject'` in build.gradle for it?

Reply:
Potentially, though I think there could be cases where they're defining it for unrelated reasons or maybe even using Kotlin instead of Groovy DSL. We could actually check this for sure from within a Gradle plugin, but it would be difficult to wire in one to just get that info here in the middle of a Dart function.

Reply:
Right - I'm thinking wrap the existing check in a check for subprojects.  If we find the "bad" text, we replace it.  If not, we know there might be subprojects we're not fixing and we warn.

Reply:
Resolved offline. `gradle.dart` depends on having a Groovy `build.gradle` already, so I think grepping the file and warning when we detect a projects block that we can't automatically migrate is sufficient here.